lib/ClangImporter/ImportName.cpp

```diff
@@ -1745,11 +1745,10 @@ ImportedName NameImporter::importNameImpl(const clang::NamedDecl *D,
       break;
     case clang::OverloadedOperatorKind::OO_Subscript: {
       auto returnType = functionDecl->getReturnType();
-      if (!returnType->isReferenceType()) {
-        // TODO: support non-reference return types (SR-14351)
-        return ImportedName();
-      }
-      if (returnType->getPointeeType().isConstQualified()) {
+      if ((!returnType->isReferenceType() && !returnType->isAnyPointerType()) ||
```

Comment:
Re our chat about this still crashing: we don't actually want to generate a setter if this returns a pointer. So, I think this should just be `if (!returnType->isReferenceType() || isConst)`. Because, even though its a "pointer" `UnsafeMutablePointer` still has value semantics, so we can't do `ValueType[0] = ptr`. Does that make sense?

Reply:
Not generating a setter in this case makes sense to me, however, `if (!returnType->isReferenceType() || isConst)` won't work well if both `const p* operator[]` & `p* operator[]` are defined, since the condition would be satisfied for both of them, and we'd generate an accessor from the overload that just happened to be defined the last.

I think a better approach would be to generate a subscript getter returning

* `UnsafeMutablePointer` if an `p* operator[]` is defined, ignoring the non-const overload
* `UnsafePointer` if only `const p* operator[]` is defined

WDYT?

Reply:
That sounds good to me :)

Reply:
@egorzhdan @zoecarver Ah yeah, I agree. Last night I had posted a response (which I took down BTW), thinking things were working but I had only been running the IDE tool. 

So, I tried a couple of things. I tried force unwrapping the `__operatorSubscript` and the `subscript` to the point where you are getting a value out, but I didn't like that approach at all. I also tried making the non-const pointer case generate a non-const `__operatorSubscript` getter and that seems to work really nicely:

```
      } else if (returnType->isAnyPointerType()) {
        baseName = "__operatorSubscript";
        result.info.accessorKind = ImportedAccessorKind::SubscriptGetter;
      } 
```

One thing, as @zoecarver mentioned is the implicit unwrap to make both the `__operatorSubscript` and the `subscript` have matching optional type pf `UnsafeMutablePointer<Int32>!`. Is this something we still want to include? 

Me and Zoe spoke on slack about the `foo[0] = ptr` case as well. I think it makes sense to not support this. 

I will be posting an updated diff very shortly.

Reply:
BTW, @egorzhdan and @zoecarver Thanks for the guidance!

Reply:
> That sounds good to me :)

Actually, to elaborate a bit more. I'd like to point out that this is a more broad problem that we're going to have to solve. There are _lots_ of places in C++ where two overloads are defined for the same function/operator, one const and one not. This means there are several places where the overload is ambiguous based on constness. I think at some point we should discuss the best _general_ solution to proceed on the forums. 

> One thing, as @zoecarver mentioned is the implicit unwrap to make both the __operatorSubscript and the subscript have matching optional type pf UnsafeMutablePointer<Int32>!. Is this something we still want to include?

Yes, this is still 100% necessary. Just to reiterate from my last comment, in any case, both `__operatorSubscript` and the subscript operator _have the same return type_. The question is whether they have the same "is conditionally unwrapped" policy, and we need that to be the case, for correctness (regardless of whatever we do with pointers).

Reply:
I did the implicit unwrap to have the optional types match. Resolving.